tests/unit_tests/utils/test_hash_generation_for_object.py

```diff
@@ -0,0 +1,158 @@
+    test2 = ObjectWithMutlipleAttributes(test1=HashableTestObject(1), test2=HashableTestObject(2))
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 == result2
+
+
+def test_object_with_multiple_attributes_not_equal():
+    test1 = ObjectWithMutlipleAttributes(test1=HashableTestObject(1), test2=HashableTestObject(2))
+    test2 = ObjectWithMutlipleAttributes(test1=HashableTestObject(1), test2=HashableTestObject(3))
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 != result2
+
+
+@dataclasses.dataclass
+class ObjectWithDict:
+    test: Dict[HashableTestObject, HashableTestObject]
+
+
+def test_object_with_dict_equal():
+    test1 = ObjectWithDict(test={HashableTestObject(1): HashableTestObject(1),
+                                 HashableTestObject(2): HashableTestObject(2)})
+    test2 = ObjectWithDict(test={HashableTestObject(1): HashableTestObject(1),
+                                 HashableTestObject(2): HashableTestObject(2)})
+
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 == result2
+
+
+def test_object_with_dict_different_values_not_equal():
+    test1 = ObjectWithDict(test={HashableTestObject(1): HashableTestObject(1),
+                                 HashableTestObject(2): HashableTestObject(2)})
+    test2 = ObjectWithDict(test={HashableTestObject(1): HashableTestObject(1)})
+
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 != result2
+
+
+def test_object_with_dict_different_order_not_equal():
+    test1 = ObjectWithDict(test={HashableTestObject(1): HashableTestObject(1),
+                                 HashableTestObject(2): HashableTestObject(2)})
+    test2 = ObjectWithDict(test={HashableTestObject(2): HashableTestObject(2),
+                                 HashableTestObject(1): HashableTestObject(1)})
+
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 != result2
+
+
+class NotHashable:
+    __hash__ = None
+
+    def __init__(self, i: int):
+        self.i = i
+
+
+@dataclasses.dataclass
+class ObjectWithNotHashable:
+    test: NotHashable
+
+
+def test_object_not_hashable_same_value_equal_hash():
+    test1 = ObjectWithNotHashable(test=NotHashable(1))
+    test2 = ObjectWithNotHashable(test=NotHashable(1))
+
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 == result2
+
+
+def test_object_not_hashable_different_values_equal_hash():
+    test1 = ObjectWithNotHashable(test=NotHashable(1))
+    test2 = ObjectWithNotHashable(test=NotHashable(2))
+
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 == result2
+
+
+@dataclasses.dataclass
+class ObjectWithCycle:
```

Comment:
Actually, I think ObjectWithDuplicates would be a better name

Reply:
I choese this name, because the test this class corresponds to , is about cycles

Reply:
Ok that's fine